server/src/grpc/storage_service/sql_query.rs

```diff
@@ -0,0 +1,430 @@
+                msg: "Query timeout",
+            }
+            .fail();
+        }
+    }
+
+    // Execute in interpreter
+    let interpreter_ctx = InterpreterContext::builder(request_id, deadline)
+        // Use current ctx's catalog and schema as default catalog and schema
+        .default_catalog_and_schema(ctx.catalog().to_string(), ctx.schema().to_string())
+        .build();
+    let interpreter_factory = Factory::new(
+        instance.query_executor.clone(),
+        instance.catalog_manager.clone(),
+        instance.table_engine.clone(),
+        instance.table_manipulator.clone(),
+    );
+    let interpreter = interpreter_factory
+        .create(interpreter_ctx, plan)
+        .map_err(|e| Box::new(e) as _)
+        .with_context(|| ErrWithCause {
+            code: StatusCode::INTERNAL_SERVER_ERROR,
+            msg: "Failed to create interpreter",
+        })?;
+
+    let output = if let Some(deadline) = deadline {
+        tokio::time::timeout_at(
+            tokio::time::Instant::from_std(deadline),
+            interpreter.execute(),
+        )
+        .await
+        .map_err(|e| Box::new(e) as _)
+        .context(ErrWithCause {
+            code: StatusCode::REQUEST_TIMEOUT,
+            msg: "Query timeout",
+        })?
+    } else {
+        interpreter.execute().await
+    }
+    .map_err(|e| Box::new(e) as _)
+    .with_context(|| ErrWithCause {
+        code: StatusCode::INTERNAL_SERVER_ERROR,
+        msg: format!("Failed to execute interpreter, sql:{}", req.sql),
+    })?;
+
+    info!(
+        "Grpc handle query success, catalog:{}, schema:{}, request_id:{}, cost:{}ms, request:{:?}",
+        ctx.catalog(),
+        ctx.schema(),
+        request_id,
+        begin_instant.saturating_elapsed().as_millis(),
+        req,
+    );
+
+    Ok(output)
+}
+
+// TODO(chenxiang): Output can have both `rows` and `affected_rows`
+fn convert_output(
+    output: &Output,
+    min_rows_per_query_batch: usize,
+    query_response_size_compression_threshold: usize,
+) -> Result<SqlQueryResponse> {
+    match output {
+        Output::Records(batches) => {
+            let mut writer = QueryResponseWriter::new(
+                min_rows_per_query_batch,
+                query_response_size_compression_threshold,
+            );
+            writer.write_batches(batches)?;
+            writer.finish()
+        }
+        Output::AffectedRows(rows) => {
+            Ok(QueryResponseBuilder::with_ok_header().build_with_affected_rows(*rows))
+        }
+    }
+}
+
+/// Writer for encoding multiple [`RecordBatch`]es to the [`SqlQueryResponse`].
+///
+/// Multiple record batches may be encoded into one batch in the query response
+/// to ensure the one batch contains at least `min_rows_per_query_batch`
+/// records.
+///
+/// Whether to do compression depends on the size of the encoded bytes.
+///
+/// REQUIRE: Multiple record batches must share the same schema.
+pub struct QueryResponseWriter {
```

Comment:
Add some basic test for this Writer, the logic inside it is a little complex.

Reply:
After refactoring, this logic inside the Writer becomes very simple.